ulib/FStar.Monotonic.HyperStack.fst

```diff
@@ -258,7 +258,7 @@ let lemma_sel_same_addr' (#a:Type0) (#rel:preorder a) (h:mem) (r1:mreference a r
          (ensures  (h `contains` r2 /\ sel h r1 == sel h r2))
 	 [SMTPatOr [
            [SMTPat (sel h r1); SMTPat (sel h r2)];
-           [SMTPat (frameOf r1); SMTPat (frameOf r2); SMTPat (as_addr r1); SMTPat (as_addr r2)]
+           [SMTPat (frameOf r1); SMTPat (frameOf r2); SMTPat (as_addr r1); SMTPat (as_addr r2); SMTPat(h `contains` r1)]
```

Comment:
The change seems a bit ad hoc. 

 -- Why only this branch of the disjunction? Why not thee previous one too?
 -- Why only ```h `contains` r1``` why not also add a disjunct for ```h `contains` r2```?

@aseemr  should comment on this too.

Reply:
Yep, this is totally ad hoc, for the sole purpose of mentioning `h` in this branch too, so that Z3 doesn't complain about the pattern not containing all variables (thusly confusing users). Because of that, it probably never used the pattern, so there's a tiny chance that it will now start to use it (and waste time). I only added the one term from the pre-condition, but it may well be a good idea to add the other one too, but you know much better than me how these things will behave when they are used.

Reply:
In the master branch, the pattern is not a disjunction, rather it is just `[SMTPat (sel h r1); SMTPat (sel h r2)]`. I am guessing the disjunction was added because the lemma is providing two properties: one that `sel` is same, and the other that `contains` also holds for the second reference. And you wanted the lemma to fire to prove the `contains` property, without having `sel` terms in the context.  Is that the disjunction was added?

To me it seems like the second branch could just be: `[SMTPat (contains h r1); SMTPat (contains h r2)]`, the other conjuncts `frameOf` and `addr_of` look seem too general.

Another option is to split the lemma into two: one that provides the `contains` property, and one that provides `sel`. And then the patterns of both can be simple non-OR patterns. Not sure which style is preferable.

FWIW, I looked in the heap, and there too the two properties are provided in the same lemma, with just `sel` terms in the pattern. That seems suboptimal too.